File: shared/api/analysis.ts
import { invoke } from '@tauri-apps/api/core'
import type { Track } from '../types'

/**
 * Analyze tracks for BPM and key detection
 * Progress events are emitted via 'analysis-track-event' Tauri event for each track
 * `force` defaults to false and bypasses the skip guard for tracks that already have BPM/key
 */
export async function analyzeTracks(trackIds: string[], force: boolean = false): Promise<void> {
	return invoke('analyze_tracks', { trackIds, force })
}

/**
 * Cancel analysis for a specific track
 * Returns true if the track was found and cancelled, false otherwise
 */
export async function cancelTrackAnalysis(trackId: string): Promise<boolean> {
	return invoke<boolean>('cancel_track_analysis', { trackId })
}

/**
 * Cancel all running analysis operations (legacy)
 */
export async function cancelAnalysis(): Promise<void> {
	return invoke('cancel_analysis')
}

/**
 * Get updated tracks after analysis
 */
export async function getAnalyzedTracks(trackIds: string[]): Promise<Track[]> {
	return invoke<Track[]>('get_analyzed_tracks', { trackIds })
}

File: src-tauri/src/commands/analysis.rs
use tauri::{AppHandle, State};

use crate::error::Result;
use crate::models::Track;
use crate::services::AnalysisService;

/// Analyze tracks for BPM and key detection with per-track events
#[tauri::command]
pub async fn analyze_tracks(
    track_ids: Vec<String>,
    analysis: State<'_, AnalysisService>,
    app_handle: AppHandle,
    force: bool,
) -> Result<()> {
    analysis
        .analyze_tracks_async(app_handle, track_ids, force)
        .await
}

/// Cancel analysis for a specific track
#[tauri::command]
pub async fn cancel_track_analysis(
    track_id: String,
    analysis: State<'_, AnalysisService>,
) -> Result<bool> {
    analysis.cancel_track_analysis(&track_id)
}

/// Cancel all running analysis operations (legacy)
#[tauri::command]
pub async fn cancel_analysis(analysis: State<'_, AnalysisService>) -> Result<()> {
    analysis.cancel_analysis()
}

/// Get updated tracks after analysis
#[tauri::command]
pub async fn get_analyzed_tracks(
    track_ids: Vec<String>,
    analysis: State<'_, AnalysisService>,
) -> Result<Vec<Track>> {
    let mut tracks = Vec::new();
    for id in track_ids {
        let track = analysis.get_updated_track(&id)?;
        tracks.push(track);
    }
    Ok(tracks)
}
